Security + Jwt 인증 로직 리팩토링

seonpilKim opened the pull request:
<!--
✅ Resolve: #이슈번호 형태로 입력해 주세요.
ex) Resolve: #123
-->
## 📌 Linked Issues
- Resolve: #13 

<!--
✅ 변경 사항을 자세히 알려주세요. (why -> what -> how)
-->
## 🔎 Change Details
### 1. Security + Jwt 인증 로직 리팩토링
#### JwtAuthenticationFilter 추가
```java
public class JwtAuthenticationFilter extends AbstractAuthenticationProcessingFilter {
...
	@Override
	public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response) throws
		AuthenticationException {
		final String authorizationHeader = HttpServletRequestHeaderUtils.extractAuthorizationHeader(request);
		final String jwt;
		try {
			jwt = jwtUtils.extractJwt(authorizationHeader);
		} catch (InvalidRequestHeaderException e) {
			throw new JwtAuthenticationException(e.getErrors());
		}
		final JwtAuthenticationToken authentication = JwtAuthenticationToken.of(jwt);

		return super.getAuthenticationManager().authenticate(authentication);
	}
...
}
```
- attemptAuthentication() 메소드에서 헤더 추출 및 Jwt 추출 로직이 실행됩니다.
- 헤더의 값이 존재하지 않거나, 접두사가 올바르지 않은 경우 `InvalidRequestHeaderException`가 발생합니다.
    - 이를 다시 `AuthenticationException` 를 상속 받는 `JwtAuthenticationException`를 던져서 `ExceptionTranslationFilter`에서 `authenticationEntryPoint.commence()` 메소드를 호출하도록 유도하였습니다.

```java
public class JwtAuthenticationFilter extends AbstractAuthenticationProcessingFilter {
...
	@Override
	protected void successfulAuthentication(HttpServletRequest request, HttpServletResponse response, FilterChain chain,
		Authentication authResult) throws IOException, ServletException {
		SecurityContextHolder.getContext().setAuthentication(authResult);
		super.successfulAuthentication(request, response, chain, authResult);
	}

	@Override
	protected void unsuccessfulAuthentication(HttpServletRequest request, HttpServletResponse response,
		AuthenticationException failed) throws IOException, ServletException {
		SecurityContextHolder.clearContext();
		super.unsuccessfulAuthentication(request, response, failed);
	}
```
- 부모의 메소드에서 `rememberMeServices`가 동작되는 것이 불필요하다고 생각하여 간단하게 재정의하였습니다.

#### JwtAuthenticationProvider 추가
- 최종적으로 여기서 Jwt를 파싱하여 인증 객체를 리턴합니다.
- Jwt를 파싱하는 과정에서 발생할 수 있는 모든 Jwt 예외들을 여기서 catch하여 다음과 같이 처리했습니다.
    - `AuthenticationException` 를 상속 받는 `JwtAuthenticationException`를 던져서 `ExceptionTranslationFilter`에서 `authenticationEntryPoint.commence()` 메소드를 호출하도록 유도하였습니다.

### 2. 편의성 클래스 추가
#### Constants 클래스 추가(`module-common`)
- 자주 사용할만한 문자열들을 상수로 보관하는 클래스 입니다.
- 각 파일에서 리터럴로 사용되던 인자들을 모두 상수로 변경하였습니다.
- 인터페이스 상수는 Anti-pattern이라 클래스로 생성하였고, final로 선언하여 상속을 방지하였습니다. 또한, 기본생성자를 private으로 선언하여 외부에서 생성자를 호출하지 못하게 막았습니다.
    > 인터페이스 상수가 Anti-pattern인 이유 [(1) Constant_interface](https://en.wikipedia.org/wiki/Constant_interface)
    > 1. 사용하지 않는 상수 변수가 namespace를 오염시킬 수 있음
    > 2. 만약 다음 릴리즈 때, 필요 없는 상수가 있어도, 이전 버전과의 호환성을 유지하기 위해 해당 상수는 영원히 인터페이스에 남아 있어야 함
    > 3. 해당 상수가 어디에서 오는지 쉽게 볼 수 있는 IDE의 지원이 없으면 추적하는데 시간이 오래 걸림
    > 4. 인스턴스를 나타내는 인터페이스 변수는 인터페이스 이름 자체보다 유용하지 않음. (method가 없기 때문에)
    > 5. 외부 API 또는 라이브러리 형태로 제공할 때 구현 방식이 드러남
    > 6. 개발자가 인터페이스를 구현할 때 상수의 값을 변경할 위험이 있음 (컴파일 오류 발생 x)

#### HttpServletRequestHeaderUtils 클래스 추가(`module-domain`)
- 아직 여러군데서 자주 쓰일지는 미지수지만, 책임 분리를 고려해서 분리해봤습니다!

<!--
✅ 추가로 전달할 내용이 있다면 적어주세요.
-->
## 💬 Comment
- 테스트 코드는 다음 PR에서..

<!--
✅ 참고한 문서가 있다면 공유해 주세요.
-->
## 📑 References
- [(1) Constant_interface](https://en.wikipedia.org/wiki/Constant_interface)

## ✅ Check Lists
- [ ] 추가한 기능에 대한 테스트는 모두 완료하셨나요?
- [x] 코드 정렬(Ctrl + Alt + L), 불필요한 코드나 오타는 없는지 확인하셨나요?
- [x] merge할 base branch가 올바른지 확인하셨나요?

seonpilKim committed:
Feat: 불필요 파일 제거

Resolve: #13

seonpilKim committed:
Feat: 상수 클래스 추가

Resolve: #13

seonpilKim committed:
Feat: HttpServletRequestHeaderUtils 추가

- Authorization 헤더 추출 메소드 추가

Resolve: #13

seonpilKim committed:
Refactor: Security + Jwt 인증 로직 리팩토링

Resolve: #13

slolee commented on module-domain/src/main/java/com/zoopi/util/HttpServletRequestHeaderUtils.java:
```diff
@@ -0,0 +1,13 @@
+
+	public static final String AUTHORIZATION_HEADER = "Authorization";
+
+	public static String extractAuthorizationHeader(HttpServletRequest request) {
```
혹시 JwtAuthenticationFilter 에서만 사용하는데 Util 클래스로 추상화한 특별한 이유가 있는걸까요?

JwtAuthenticationFilter 의 attemptAuthentication 메소드에서 직접 request.getHeader(...) 를 호출하도록 하는건 어떤지 의견드립니다! 혹시 다른 필요성이 있다면 공유 부탁드립니다 :D

seonpilKim replied:
그러게요.. ㅋㅋㅋ 새벽에 정신없이 코딩했는데 다시보니 너무 불필요하게 분리한 것 같네요! 말씀하신대로 변경하겠습니다 :)

slolee commented on module-common/src/main/java/com/zoopi/util/Constants.java:
```diff
@@ -0,0 +1,13 @@
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
```
Private 생성자 주신거 좋은거같아요! 배워갑니다~ 👍

seonpilKim replied:
😁👍

slolee commented on module-domain/src/main/java/com/zoopi/util/JwtUtils.java:
```diff
@@ -47,27 +57,48 @@ protected void init() {
 		SECRET_KEY = Base64.getEncoder().encodeToString(SECRET_KEY.getBytes());
```
`@Value` 로 부분을 final 까지 붙여서 생성자 주입방식을 사용하는 방법도 본적있는데 참고하시면 좋을거같습니다!
`@PostConstructor`를 사용한 방법도 충분히 좋은 방법이라고 생각합니다! 선택적으로 사용해면 될거같아요 :D

```
private final String name;

public Demo(@Value("${test.name}") String name) {
    this.name = name;
}
```

seonpilKim replied:
이런 방법도 있군요..! 나중에 좀 더 알아봐야겠네요. 알려주셔서 감사합니다😄

slolee commented on module-domain/src/main/java/com/zoopi/util/JwtUtils.java:
```diff
@@ -47,27 +57,48 @@ protected void init() {
+			errors = ErrorResponse.FieldError.of(AUTHORIZATION_HEADER, EMPTY, AUTHORIZATION_HEADER_MISSING);
+		} else if (!authorizationHeader.startsWith(TOKEN_TYPE + SPACE)) {
+			errors = ErrorResponse.FieldError.of(AUTHORIZATION_HEADER, authorizationHeader, BEARER_PREFIX_MISSING);
+		} else {
```
```
if (authorizationHeader == null) {
    throw new InvalidRequestHeaderException(ErrorResponse.FieldError.of(AUTHORIZATION_HEADER, EMPTY, AUTHORIZATION_HEADER_MISSING));
} 
if (!authorizationHeader.startsWith(TOKEN_TYPE + SPACE)) {
    throw new InvalidRequestHeaderException(ErrorResponse.FieldError.of(AUTHORIZATION_HEADER, authorizationHeader, BEARER_PREFIX_MISSING));
}
```

조금 불필요한 코드들이 생기는거 같아서 이렇게 심플하게 구성하는건 어떤가 의견드려봅니다!
한줄이 너무 길어진다는 단점이 있긴 한거같습니다 ㅠ_ㅠ

seonpilKim replied:
훨씬 깔끔하네요.. 바로 적용하겠습니다 :)

dongkyunkimdev replied:
찬준님이 말씀하신 부분 관련해서 Guard Clauses 내용도 한번 스윽 보시면 좋을 것 같습니다!!
[참고](https://velog.io/@swhan9404/%ED%81%B4%EB%A6%B0%EC%BD%94%EB%93%9C%EB%A5%BC-%EC%9C%84%ED%95%9C-Guard-Clause)

seonpilKim replied:
@dongkyunkimdev 
조직이 방금 복구되어서 이제 답글 남기네요ㅠㅠ 좋은 정보 알려주셔서 감사합니다!

seonpilKim committed:
Refactor: validateAuthorizationHeader() 메소드 리팩토링

- 코드 간결화

Resolve: #13

seonpilKim committed:
Refactor: HttpServletRequestHeaderUtils 제거

- request.getHeader() 메소드 사용으로 변경
- 상수(AUTHORIZATION_HEADER) -> Constants 클래스로 위치 이동

Resolve: #13

seonpilKim committed:
Refactor: 패키지 이동

handler 패키지로 이동
- CustomAccessDeniedHandler
- CustomAuthenticationEntryPoint

Resolve: #13

seonpilKim committed:
Feat: SkipPathRequestMatcher 추가

이전의 set을 이용한 방식은 /**을 처리하지 못하여 RequestMatcher 도입

RequestMatcher를 구현하여 JWT 인증 로직이 필요 없는 paths 들은 인증 필터가 실행되지 않도록 수정

Resolve: #13

seonpilKim commented on module-domain/src/main/java/com/zoopi/config/security/matcher/SkipPathRequestMatcher.java:
```diff
@@ -0,0 +1,28 @@
+import org.springframework.security.web.util.matcher.OrRequestMatcher;
+import org.springframework.security.web.util.matcher.RequestMatcher;
+
+public class SkipPathRequestMatcher implements RequestMatcher {
```
RequestMatcher를 구현해서 인증이 필요 없는 paths에서는 인증 필터 로직을 생략할 수 있도록 코드를 추가하였습니다.
OrRequestMatcher를 이용하여 여러 Paths를 matches할 수 있게 하였습니다.
```java
public class SkipPathRequestMatcher implements RequestMatcher {

	private final OrRequestMatcher matcher;

	public SkipPathRequestMatcher(List<String> skipPaths) {
		final List<RequestMatcher> requestMatchers = skipPaths.stream()
			.map(AntPathRequestMatcher::new)
			.collect(Collectors.toList());
		this.matcher = new OrRequestMatcher(requestMatchers);
	}

	@Override
	public boolean matches(HttpServletRequest request) {
		return !matcher.matches(request);
	}

}
```

SecurityConfig에서 인증을 건너뛸 paths를 관리하고 있고
```java
private static final String[] AUTH_WHITELIST_SWAGGER = {"/v2/api-docs", "/configuration/ui",
		"/swagger-resources/**", "/configuration/security", "/swagger-ui.html/**", 
		"/webjars/**", "/swagger/**"};
private static final String[] AUTH_WHITELIST_GUEST = {"/", "/csrf", "/error", "/auth/**", "/oauth2/**"};
```

JwtAuthenticationFilter에서 requiresAuthentication 메소드를 재정의하였습니다.
```java
public class JwtAuthenticationFilter extends AbstractAuthenticationProcessingFilter {

	private final RequestMatcher matcher;

	public JwtAuthenticationFilter(RequestMatcher matcher, JwtUtils jwtUtils) {
		super(ALL_PATTERN);
		this.jwtUtils = jwtUtils;
		this.matcher = matcher;
	}

	@Override
	protected boolean requiresAuthentication(HttpServletRequest request, HttpServletResponse response) {
		return matcher.matches(request);
	}
...
}
```

SecurityConfig에서 정의한 skipPaths들은 인증 필터가 동작하지 않는 것을 확인하였습니다.

dongkyunkimdev commented on module-common/src/main/java/com/zoopi/util/Constants.java:
```diff
@@ -0,0 +1,13 @@
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class Constants {
```
👍

dongkyunkimdev commented:
잘 작성하신 것 같아서 코멘트가 별로 없네요,,  고생하셨습니다!!👍👍
